pkg/collector/corechecks/servicediscovery/module/impl_linux.go

```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+	err = kernel.WithNS(ns, func() error {
+		addrs, err := net.InterfaceAddrs()
+		if err != nil {
+			return nil
+		}
+		for _, addr := range addrs {
+			if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+				if ipnet.IP.To4() != nil {
+					out = append(out, ipnet.IP.String())
+				}
+			}
+		}
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// Lifted from pkg/network/proc_net.go
+const (
+	tcpListen uint64 = 10
+
+	// tcpClose is also used to indicate a UDP connection where the other end hasn't been established
+	tcpClose  uint64 = 7
+	udpListen        = tcpClose
+)
+
+// Based on snapshotBoundSockets() in
+// pkg/security/security_profile/activity_tree/process_node_snapshot.go.  Get
+// the list of open ports with socket inodes for the provided namespace.
+func getNsInfo(pid int, isRoot bool) (*namespaceInfo, error) {
+	path := filepath.Join(kernel.HostProc(fmt.Sprintf("%d", pid)))
+	proc, err := procfs.NewFS(path)
+	if err != nil {
+		log.Warnf("error while opening procfs (pid: %v): %s", pid, err)
+		return nil, err
+	}
+
+	TCP, err := proc.NetTCP()
+	if err != nil {
+		log.Debugf("couldn't snapshot TCP sockets: %v", err)
+	}
+	UDP, err := proc.NetUDP()
+	if err != nil {
+		log.Debugf("couldn't snapshot UDP sockets: %v", err)
+	}
+	TCP6, err := proc.NetTCP6()
+	if err != nil {
+		log.Debugf("couldn't snapshot TCP6 sockets: %v", err)
+	}
+	UDP6, err := proc.NetUDP6()
+	if err != nil {
+		log.Debugf("couldn't snapshot UDP6 sockets: %v", err)
+	}
+
+	nsInfo := namespaceInfo{
+		socketInodeToPort: make(map[uint64]int),
+	}
+
+	for _, sock := range TCP {
+		if sock.St != tcpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range TCP6 {
+		if sock.St != tcpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range UDP {
+		if sock.St != udpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range UDP6 {
+		if sock.St != udpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+
+	if !isRoot {
+		nsInfo.addrs, err = getNsAddresses(path)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &nsInfo, nil
+}
+
+func getRootNsInode(procRoot string) (uint32, error) {
+	ns, err := kernel.GetRootNetNamespace(procRoot)
+	if err != nil {
+		return 0, err
+	}
+	defer ns.Close()
+
+	inode, err := kernel.GetInoForNs(ns)
+	if err != nil {
+		return 0, err
+	}
+
+	return inode, nil
+}
+
+// parsingContext holds temporary context not preserved between invocations of
+// the endpoint.
+type parsingContext struct {
+	procRoot  string
+	rootNs    uint32
+	netNsInfo map[uint32]*namespaceInfo
+}
+
+func (s *discovery) getListener(context parsingContext, pid int32) *model.Listener {
+	proc, err := process.NewProcess(pid)
+	if err != nil {
+		return nil
+	}
+
+	name, err := proc.Name()
```

Comment:
this is wrong, it returns the process name, instead of the service name

Reply:
The service name was being calculated in the agent check but now it's moved here.